spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkPushedDownAggregateUtil.java

```diff
@@ -0,0 +1,376 @@
+    StructType finalSchema = new StructType();
+    for (int index = 0; index < aggregates.size(); index++) {
+      if ((aggregates.get(index)).op().name().equals("COUNTSTAR")) {
+        StructField field =
+            new StructField("COUNT(*)", DataTypes.LongType, false, Metadata.empty());
+        finalSchema = finalSchema.add(field);
+      } else {
+        String colName = ((UnboundAggregate) aggregates.get(index)).ref().name();
+        DataType dataType = getDataTypeForAggregateColumn(colName, caseSensitive, schema);
+        if (dataType instanceof StructType
+            || dataType instanceof ArrayType
+            || dataType instanceof MapType) {
+          // not building pushed down aggregate schema for complex types to disable aggregate push
+          // down because the statistic info for complex are not available.
+          return finalSchema;
+        }
+        if ((aggregates.get(index)).op().name().equals("COUNT")) {
+          StructField field =
+              new StructField(
+                  "COUNT(" + colName + ")", DataTypes.LongType, false, Metadata.empty());
+          finalSchema = finalSchema.add(field);
+        } else if ((aggregates.get(index)).op().name().equals("MAX")) {
+          StructField field =
+              new StructField("MAX(" + colName + ")", dataType, false, Metadata.empty());
+          finalSchema = finalSchema.add(field);
+        } else if ((aggregates.get(index)).op().name().equals("MIN")) {
+          StructField field =
+              new StructField("MIN(" + colName + ")", dataType, false, Metadata.empty());
+          finalSchema = finalSchema.add(field);
+        }
+      }
+    }
+    return finalSchema;
+  }
+
+  public static DataType getDataTypeForAggregateColumn(
+      String colName, boolean caseSensitive, Schema schema) {
+    Type type = null;
+    for (int i = 0; i < schema.columns().size(); i++) {
+      if ((caseSensitive && schema.columns().get(i).name().equals(colName))
+          || (!caseSensitive && schema.columns().get(i).name().equalsIgnoreCase(colName))) {
+        type = schema.columns().get(i).type();
+      }
+    }
+    return SparkSchemaUtil.convert(type);
+  }
+
+  /**
+   * get "lower_bounds", "upper_bounds", "record_count", "null_value_counts" from metadata table and
+   * use these to calculate Max/Min/Count, and then use the values of Max/Min/Count to construct an
+   * InternalRow to return to Spark.
+   */
+  @SuppressWarnings("checkstyle:CyclomaticComplexity")
+  public static InternalRow[] constructInternalRowForPushedDownAggregate(
+      SparkSession spark, Table table, StructType pushedAggregateSchema, boolean caseSensitive) {
+    // get the statistics info from DATA_FILES, calculate the aggregate values (min/max/count)
+    // and use these value to build an internalRow.
+    Dataset<Row> metadataRows =
+        SparkTableUtil.loadMetadataTable(spark, table, MetadataTableType.DATA_FILES);
```

Comment:
I think there should be a flag to enable/disable running a parallel operation from within job planning. Nested a DataFrame operation like this does not seem like a good idea in all cases. I can certainly see why we want to do this, but I think we should have an implementation that uses `planFiles` first and then add this as an optimization.